## Establish Dependencies and Load in Data

In [1]:
# Import all packages and dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
import sys, warnings, os
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [2]:
#Ignore warnings when validating scores
if not sys.warnoptions:
    warnings.simplefilter("ignore")
    os.environ["PYTHONWARNINGS"] = "ignore"

In [3]:
#Setting columns and rows to display all the results
pd.set_option("display.max_columns", None, "display.max_rows", None)

In [4]:
# Read in the dataset
ufc_master_df = pd.read_csv("../Resources/ufc-master.csv")
ufc_master_df.head()

R_fighter         B_fighter  R_odds  B_odds        R_ev  \
0   Alistair Overeem  Alexander Volkov     150    -182  150.000000   
1     Cory Sandhagen     Frankie Edgar    -400     300   25.000000   
2  Alexandre Pantoja        Manel Kape    -125     100   80.000000   
3     Diego Ferreira    Beneil Dariush    -125     105   80.000000   
4    Michael Johnson        Clay Guida    -225     175   44.444444   

         B_ev      date                location country Winner  title_bout  \
0   54.945055  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
1  300.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
2  100.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
3  105.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
4  175.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   

   weight_class gender  no_of_rounds  B_current_lose_streak  \
0   Heavyweight   MALE             5                      0   
1  Bantamweight   MALE             3                      0   
2     Flyweight   MALE             3                      0   
3   Lightweight   MALE             3                      0   
4   Lightweight   MALE             3                      2   

   B_current_win_streak  B_draw  B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  \
0                     1       0                  4.76               0.58   
1                     1       1                  3.70               0.39   
2                     0       0                  0.00               0.00   
3                     5       1                  3.98               0.49   
4                     0       0                  2.41               0.33   

   B_avg_SUB_ATT  B_avg_TD_landed  B_avg_TD_pct  B_longest_win_streak  \
0            0.2             0.69          0.70                     4   
1            0.3             2.28          0.31                     5   
2            0.0             0.00          0.00                     0   
3            1.1             1.75          0.32                     5   
4            0.7             3.37          0.38                     4   

   B_losses  B_total_rounds_fought  B_total_title_bouts  \
0         2                     26                    0   
1         8                     94                    9   
2         0                      0                    0   
3         4                     36                    0   
4        15                     87                    0   

   B_win_by_Decision_Majority  B_win_by_Decision_Split  \
0                           0                        1   
1                           0                        1   
2                           0                        0   
3                           0                        1   
4                           0                        3   

   B_win_by_Decision_Unanimous  B_win_by_KO/TKO  B_win_by_Submission  \
0                            2                3                    0   
1                           10                5                    2   
2                            0                0                    0   
3                            4                3                    5   
4                            8                2                    4   

   B_win_by_TKO_Doctor_Stoppage  B_wins  B_Stance  B_Height_cms  B_Reach_cms  \
0                             0       6  Orthodox        200.66       203.20   
1                             0      18  Orthodox        167.64       172.72   
2                             0       0  Southpaw        165.10       172.72   
3                             0      13  Southpaw        177.80       182.88   
4                             0      17  Orthodox        170.18       177.80   

   B_Weight_lbs  R_current_lose_streak  R_current_win_streak  R_draw  \
0           250                      0                     2       0   
1           135                      0                     1       0   
2           

In [5]:
print(ufc_master_df.shape)
print(ufc_master_df.info())

(4566, 137)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4566 entries, 0 to 4565
Columns: 137 entries, R_fighter to B_rev_bout
dtypes: bool(1), float64(73), int64(47), object(16)
memory usage: 4.7+ MB
None


## Baseline Classification Model

In [6]:
# Splitting X (features) from y (labels)
#Classification task, the target value will be "Winner"
y = ufc_master_df.Winner
X = ufc_master_df.drop(['Winner'], axis=1)

In [7]:
# Begin encoding categorical variables by separating variabes by datatype
categorical_col = [col for col in X.columns if X[col].dtypes == 'object']
numeric_col = [col for col in X.columns if col not in categorical_col]

In [8]:
# Encode categorical columns first in order to avoid data leakage in split
enc = LabelEncoder()
for i in X[categorical_col]:
    #using astype(str) to avoid columns with 'float and str' to throw errors
    X[i] = enc.fit_transform(X[i].astype(str))
    

In [9]:
# Encode label (y) for Red fighter ID = 1 Blue fighter ID = 0
y = [1 if win == "Red" else 0 for win in y]


In [10]:
# Split dataset into train test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size = 0.25)

In [11]:
# Find and fill null values
# List column key/values in descending order
X_train.isnull().sum().sort_values(ascending=False)

B_Women's Featherweight_rank    3424
R_Women's Featherweight_rank    3417
B_Pound-for-Pound_rank          3403
B_Women's Flyweight_rank        3396
R_Women's Flyweight_rank        3387
B_Women's Strawweight_rank      3386
B_Featherweight_rank            3365
B_Women's Bantamweight_rank     3365
B_Lightweight_rank              3363
B_Light Heavyweight_rank        3362
B_Bantamweight_rank             3362
B_Flyweight_rank                3358
B_Welterweight_rank             3357
B_Middleweight_rank             3356
R_Women's Strawweight_rank      3356
B_Heavyweight_rank              3346
R_Women's Bantamweight_rank     3343
R_Featherweight_rank            3333
R_Lightweight_rank              3329
R_Middleweight_rank             3329
R_Flyweight_rank                3328
R_Light Heavyweight_rank        3326
R_Bantamweight_rank             3323
R_Welterweight_rank             3317
R_Heavyweight_rank              3317
R_Pound-for-Pound_rank          3312
R_rev_bout                      1330
B

In [12]:
# Creating a StandardScaler instance.

scaler = StandardScaler()
# Fitting the Standard Scaler with the training data.
X_scaler = scaler.fit(X_train)

##Scaling the data.
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [13]:
X_train.sample(10)

R_fighter  B_fighter  R_odds  B_odds        R_ev        B_ev  date  \
1190        979        179    -210     175   47.619048  175.000000   359   
1562        512        931     215    -255  215.000000   39.215686    60   
2391        365        683     130    -150  130.000000   66.666667    11   
1236       1147        330    -160     140   62.500000  140.000000   329   
4366        621       1010    -230     190   43.478261  190.000000   128   
1145        384        895   -1000     650   10.000000  650.000000   382   
4472        225        977    -180     168   55.555556  168.000000   319   
1388        173        286    -105    -115   95.238095   86.956522   191   
2803        501        759    -260     240   38.461538  240.000000   152   
3844        940        438    -500     435   20.000000  435.000000   337   

      location  country  title_bout  weight_class  gender  no_of_rounds  \
1190        71       21       False             0       1             3   
1562        90       21       False             5       1             3   
2391        66       21       False             6       1             3   
1236        66       21        True             4       1             5   
4366        66       21        True             6       1             3   
1145       122        2       False             8       1             3   
4472        66       21       False             8       1             3   
1388        66       21       False             0       1             3   
2803       103        2       False             6       1             3   
3844        38       21       False             2       1             3   

      B_current_lose_streak  B_current_win_streak  B_draw  \
1190                      1                     0       0   
1562                      1                     0       0   
2391                      0                     2       0   
1236                      0                     1       0   
4366                      0                     0       0   
1145                      0                     0       0   
4472                      1                     0       0   
1388                      0                     2       0   
2803                      1                     0       0   
3844                      1                     0       0   

      B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  B_avg_SUB_ATT  B_avg_TD_landed  \
1190             33.250000           0.387500       0.500000         3.750000   
1562             34.000000           0.505000       0.000000         0.000000   
2391             11.500000           0.320000       0.500000         0.000000   
1236             52.818182           0.538182       0.636364         1.545455   
4366                   NaN                NaN            NaN              NaN   
1145                   NaN                NaN            NaN              NaN   
4472             28.333333           0.618333       1.500000         0.333333   
1388             80.000000           0.485000       0.000000         6.000000   
2803              6.000000           0.500000       1.000000         0.000000   
3844             32.000000           0.480000       1.000000         1.000000   

      B_avg_TD_pct  B_longest_win_streak  B_losses  B_total_rounds_fought  \
1190      0.382500                     3         1                     10   
1562      0.000000                     1         1                      4   
2391      0.000000                     2         0                      2   
1236      0.404545                     4         2                     33   
4366           NaN                     0         0                      0   
1145           NaN                     0         0                      0   
4472      0.250000                     3         2                     13   
1388      0.730000                     2         0                      6   
2803      0.000000                     0         1                      1   
3844     

In [14]:
# Dummy classifier used for simple model for comparison after optimization-DummyClassifier will allow us to run model with null values
simple_model = DummyClassifier(random_state=42)
simple_model.fit(X_train, y_train)

DummyClassifier(random_state=42)

In [15]:
prediction = simple_model.predict(X_test)
accuracy_score(y_test, prediction)

0.5175131348511384

#### Our simple model can predict winners with ~ 51.75% accuracy

In [16]:
fillna = ufc_master_df.fillna(0)

In [17]:
fillna.reset_index(level=0, inplace=True)
fillna['level_0'] = fillna['index']
fillna.head()

index          R_fighter         B_fighter  R_odds  B_odds        R_ev  \
0      0   Alistair Overeem  Alexander Volkov     150    -182  150.000000   
1      1     Cory Sandhagen     Frankie Edgar    -400     300   25.000000   
2      2  Alexandre Pantoja        Manel Kape    -125     100   80.000000   
3      3     Diego Ferreira    Beneil Dariush    -125     105   80.000000   
4      4    Michael Johnson        Clay Guida    -225     175   44.444444   

         B_ev      date                location country Winner  title_bout  \
0   54.945055  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
1  300.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
2  100.000000  2/6/2021  Las Vegas, Nevada, USA     USA    Red       False   
3  105.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   
4  175.000000  2/6/2021  Las Vegas, Nevada, USA     USA   Blue       False   

   weight_class gender  no_of_rounds  B_current_lose_streak  \
0   Heavyweight   MALE             5                      0   
1  Bantamweight   MALE             3                      0   
2     Flyweight   MALE             3                      0   
3   Lightweight   MALE             3                      0   
4   Lightweight   MALE             3                      2   

   B_current_win_streak  B_draw  B_avg_SIG_STR_landed  B_avg_SIG_STR_pct  \
0                     1       0                  4.76               0.58   
1                     1       1                  3.70               0.39   
2                     0       0                  0.00               0.00   
3                     5       1                  3.98               0.49   
4                     0       0                  2.41               0.33   

   B_avg_SUB_ATT  B_avg_TD_landed  B_avg_TD_pct  B_longest_win_streak  \
0            0.2             0.69          0.70                     4   
1            0.3             2.28          0.31                     5   
2            0.0             0.00          0.00                     0   
3            1.1             1.75          0.32                     5   
4            0.7             3.37          0.38                     4   

   B_losses  B_total_rounds_fought  B_total_title_bouts  \
0         2                     26                    0   
1         8                     94                    9   
2         0                      0                    0   
3         4                     36                    0   
4        15                     87                    0   

   B_win_by_Decision_Majority  B_win_by_Decision_Split  \
0                           0                        1   
1                           0                        1   
2                           0                        0   
3                           0                        1   
4                           0                        3   

   B_win_by_Decision_Unanimous  B_win_by_KO/TKO  B_win_by_Submission  \
0                            2                3                    0   
1                           10                5                    2   
2                            0                0                    0   
3                            4                3                    5   
4                            8                2                    4   

   B_win_by_TKO_Doctor_Stoppage  B_wins  B_Stance  B_Height_cms  B_Reach_cms  \
0                             0       6  Orthodox        200.66       203.20   
1                             0      18  Orthodox        167.64       172.72   
2                             0       0  Southpaw        165.10       172.72   
3                             0      13  Southpaw        177.80       182.88   
4                             0      17  Orthodox        170.18       177.80   

   B_Weight_lbs  R_current_lose_streak  R_current_win_streak  R_draw  \
0           250                      0                     2       0   
1           135                   

In [18]:
# Saving csv with null values filled
file_path = "../Resources/na_filled.csv"
fillna.to_csv(file_path, index=False, header=True)